### DISCUSSION TITLE
Multiple sites on one server. One instance of pocketbase, or many?

### DISCUSSION THREAD

**Discussion by oscadev** [pocketbase/pocketbase#738] [Q&A] [answered]
I want to host many sites on a single server, and have each site have its own login to CRUD stuff for that site with pocketbase. Should I  run a separate instance of pocketbase for each site on a different port, or am I supposed to somehow isolate users to their own models/dbs in one instance of pocketbase?

**Comment by ganigeorgiev** [comment, marked as answer]
It depends on your use case.

You can do both, but I'd usually recommend the first approach if your sites are not related, aka. multiple PocketBase processes each on different port and their own `pb_data` directory because it would be a lot easier to isolate your data this way:

```sh
./pocketbase serve --dir="./pb_data1" --http="127.0.0.1:8091" & \
./pocketbase serve --dir="./pb_data2" --http="127.0.0.1:8092" &

# to kill all instances:
sudo killall pocketbase
```
You could put all instances behind a reverse proxy like nginx/apache/caddy/etc. and access them by location path or unique domain/subdomain. Optionally, you can also have a "supervisor" instance that controls the other ones (I think @benallfree uses a similar approach for his hosting service https://github.com/benallfree/pockethost).

If you want to go with a single instance for everything, you can isolate your data using the collection API rules (they act both as access rules and filters). The documentation currently is a bit lacking, but for more info you could check - https://pocketbase.io/docs/manage-collections/#collection-api-rules.

**Reply by oscadev** [reply]
Thanks, seems like just having a new instance for each isolated project makes the most sense.

**Reply by benallfree** [reply]
> If you want to go with a single instance for everything, you can isolate your data using the collection API rules (they act both as access rules and filters)

@ganigeorgiev  Are you thinking that each user would have a profile field indicating which tenant they belong to, and then each collection has a `tenantId` or something?

**Reply by ganigeorgiev** [reply]
@benallfree Yes, but this requires very carefully to construct your API rules and may get complicated really quickly if you have a lot of collections and this is why I wouldn't recommend it.

**Reply by benallfree** [reply]
@ganigeorgiev Im curious…other than routes and sqlite connections, what other considerations can you think of for multitenant support?

**Reply by ganigeorgiev** [reply]
@benallfree Multi-tenant, as in having multiple separate databases for each tenant, for now it is not planned. I think it will complicate the API and the collections management in general. And I'm sure that at some point there will be cases where someone will want to link/share data between the databases and it can become really messy.

If you have different non-related applications, it's better to keep them separate. This way also it will be a lot easier to move the tenant to a different server when a single shared vm is no longer sufficient.

**Reply by nileshtrivedi** [reply]
I have a use-cases where I'd prefer to run a single pocketbase process for multiple isolated databases. Each db belongs to a different project and will have different auth token, so no data sharing/linking is ever expected. But it is crucial to keep the cost of APIs drastically low.

I might take a stab at implementing this in a fork. Would be great if someone were to outline the steps.

**Reply by nileshtrivedi** [reply]
FWIW, I [requested this feature in sqlite-rest](https://github.com/b4fun/sqlite-rest/issues/108) as well.

**Comment by benallfree** [comment]
Hello, just chiming in to say that yes currently Pockethost.io uses a supervisor (written in nodejs) that launches and maintains separate processes - one per subdomain. The supervisor accesses a supervisor instance of pocketbase where all the subdomain mappings and PIDs are tracked.

The next evolution of Pockethost might include use Firecracker on fly.io to maintain instances just-in-time, or perhaps we will even  accept fly.io machine credentials and launch instances under separate fly.io accounts. That option would be for larger deployments where Pockethost was just being used as a management console more purpose-built for PocketBase.

Having said all that, I continue to be curious about a multitenant version of PocketBase. It would be a significant overhaul though. We would be to look for an `X-PocketBase` request header and switch SQLite connections and API routes based on that. Go is not my strong suit, but I may attempt it at some point because I think there is a strong use case for a instantly provisioning tens of thousands of low-traffic tenants. Especially since interest in PocketBase is growing and people want to try it quickly.

**Comment by adampatterson** [comment]
Is it possible to specify the DB file name in the service? 

I have 2 services running and would like to keep them separated.

**Reply by ganigeorgiev** [reply]
You can specify a different data directory (the db file is stored inside it) using the `--dir` flag as shown in the comment above - https://github.com/pocketbase/pocketbase/discussions/738#discussioncomment-3826158:

```sh
./pocketbase serve --dir="/path/to/your/pb_data"
```

You can inspect all available flags with `--help`.

**Reply by adampatterson** [reply]
Oh, so like this `ExecStart      = /home/adam/pb/pocketbase serve --http="ip:port" --dir="/path/to/your/pb_data"`

Thanks!

**Reply by adampatterson** [reply]
Last question :) 

Do the migrations end up in a separate folder?

I am attempting to move a test server into another server that now has 2 instances running.

**Reply by ganigeorgiev** [reply]
By default the `pb_migrations` directory is expected to be in the same parent directory as the `pb_data` (aka. /pb_data/../pb_migrations).

If you want your instances to use the same `pb_migrations`, you can specify the `--migrationsDir` flag.

**Reply by adampatterson** [reply]
I'd like to use different migration folders ( I assume if my migrations are in one folder each service would end up with the same schema or is that incorrect?

**Reply by ganigeorgiev** [reply]
> I assume if my migrations are in one folder each service would end up with the same schema or is that incorrect?

Yes, if your services use the same migrations folder then they will have the same collections schema.

If you want to use different migration folders then you can explicitly specify the directory paths via the `--migrationsDir` flag for each service OR leave the default and only specify `--dir` (because by default it is relative to the `pb_data` location, aka. `../pb_migrations`).

**Reply by adampatterson** [reply]
Thanks so much!

**Comment by izulwahidin** [comment]
Check this out [PocketBase Multi Instance](https://github.com/izulwahidin/Docker-PocketBase-Multi-Instance)
